### PR TITLE
avoid accesses to instructiont::guard

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5779e2aa0980cfaecacf6fe4bacd42273ab63c2c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100335530

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f8538c2a289ac6ac776f27886f299ecad47d5cf7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100338270

File: src/goto-programs/adjust_float_expressions.cpp [hunk 1]
```diff
@@ -200,7 +200,9 @@ void adjust_float_expressions(
   Forall_goto_program_instructions(it, goto_function.body)
   {
     adjust_float_expressions(it->code, ns);
-    adjust_float_expressions(it->guard, ns);
+
+    if(it->has_condition())
+      adjust_float_expressions(it->guard, ns);
```
Comment (tautschnig) [inline src/goto-programs/adjust_float_expressions.cpp]: Add `set_condition`?

File: src/goto-programs/goto_convert_functions.cpp [hunk 2]
```diff
@@ -109,9 +109,12 @@ void goto_convert_functionst::add_return(
     while(true)
     {
       // unconditional goto, say from while(1)?
-      if(last_instruction->is_goto() &&
-         last_instruction->guard.is_true())
+      if(
+        last_instruction->is_goto() &&
+        last_instruction->get_condition().is_true())
```
Comment (tautschnig) [inline src/goto-programs/goto_convert_functions.cpp:114]: Change in wrong commit

File: src/goto-programs/goto_inline_class.cpp [hunk 3]
```diff
@@ -315,8 +315,10 @@ void goto_inlinet::insert_function_body(
     Forall_goto_program_instructions(it, tmp)
     {
       replace_location(it->source_location, target->source_location);
-      replace_location(it->guard, target->source_location);
       replace_location(it->code, target->source_location);
+
+      if(it->has_condition())
+        replace_location(it->guard, target->source_location);
```
Comment (tautschnig) [inline src/goto-programs/goto_inline_class.cpp]: Add `set_condition`?

File: src/goto-programs/link_goto_model.cpp [hunk 4]
```diff
@@ -33,7 +33,9 @@ static void rename_symbols_in_function(
   Forall_goto_program_instructions(iit, program)
   {
     rename_symbol(iit->code);
-    rename_symbol(iit->guard);
+
+    if(iit->has_condition())
+      rename_symbol(iit->guard);
```
Comment (tautschnig) [inline src/goto-programs/link_goto_model.cpp]: Add `set_condition`?

File: src/goto-programs/link_goto_model.cpp [hunk 5]
```diff
@@ -144,7 +146,9 @@ static bool link_functions(
       Forall_goto_program_instructions(iit, dest_it->second.body)
       {
         object_type_updates(iit->code);
-        object_type_updates(iit->guard);
+
+        if(iit->has_condition())
+          object_type_updates(iit->guard);
```
Comment (tautschnig) [inline src/goto-programs/link_goto_model.cpp:155]: Add `set_condition`?

File: src/goto-programs/remove_complex.cpp [hunk 6]
```diff
@@ -288,7 +288,9 @@ static void remove_complex(
   Forall_goto_program_instructions(it, goto_function.body)
   {
     remove_complex(it->code);
-    remove_complex(it->guard);
+
+    if(it->has_condition())
+      remove_complex(it->guard);
```
Comment (tautschnig) [inline src/goto-programs/remove_complex.cpp]: Add `set_condition`?

File: src/goto-programs/remove_vector.cpp [hunk 7]
```diff
@@ -221,7 +221,9 @@ void remove_vector(goto_functionst::goto_functiont &goto_function)
   Forall_goto_program_instructions(it, goto_function.body)
   {
     remove_vector(it->code);
-    remove_vector(it->guard);
+
+    if(it->has_condition())
+      remove_vector(it->guard);
```
Comment (tautschnig) [inline src/goto-programs/remove_vector.cpp]: Add `set_condition`?

File: src/goto-programs/rewrite_union.cpp [hunk 8]
```diff
@@ -104,7 +104,9 @@ void rewrite_union(goto_functionst::goto_functiont &goto_function)
   Forall_goto_program_instructions(it, goto_function.body)
   {
     rewrite_union(it->code);
-    rewrite_union(it->guard);
+
+    if(it->has_condition())
+      rewrite_union(it->guard);
```
Comment (tautschnig) [inline src/goto-programs/rewrite_union.cpp]: Add `set_condition`?

File: src/goto-programs/slice_global_inits.cpp [hunk 9]
```diff
@@ -67,8 +67,12 @@ void slice_global_inits(goto_modelt &goto_model)
     {
       const codet &code = i_it->code;
       find_symbols(code, symbols, true, false);
-      const exprt &expr = i_it->guard;
-      find_symbols(expr, symbols, true, false);
+
+      if(i_it->has_condition())
+      {
+        const exprt &expr = i_it->get_condition();
```
Comment (tautschnig) [inline src/goto-programs/slice_global_inits.cpp]: Maybe we can just get rid of this temporary that doesn't actually add value?

File: src/pointer-analysis/goto_program_dereference.cpp [hunk 10]
```diff
@@ -301,7 +301,8 @@ void goto_program_dereferencet::dereference_instruction(
   #endif
   goto_programt::instructiont &i=*target;
 
-  dereference_expr(i.guard, checks_only, value_set_dereferencet::modet::READ);
+  if(i.has_condition())
+    dereference_expr(i.guard, checks_only, value_set_dereferencet::modet::READ);
```
Comment (tautschnig) [inline src/pointer-analysis/goto_program_dereference.cpp]: Add `set_condition`?

**Review by tautschnig — APPROVED** [review]

**Review by smowton — APPROVED** [review]
Looks good except as per @tautschnig you should probably add some non-const accessor for the guard

File: src/goto-programs/goto_convert_functions.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-programs/goto_convert_functions.cpp:114]: fixed

**Comment by kroening** [comment]
@smowton @tautschnig I couldn't decide between non-const `get_condition()` or a `set_condition(...)`. It probably makes sense to do the non-const one, in like with the other get_() methods.

File: src/goto-programs/slice_global_inits.cpp (hunk 9 above)
Comment (kroening) [reply src/goto-programs/slice_global_inits.cpp]: done

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f34ad060f9ff89cc5340b99c83f2a333c9d179b4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100527912
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 17ceafd074f81ba3455f7f28df601cf56e652493).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100530206

File: src/goto-programs/goto_program.h [hunk 11]
```diff
@@ -287,6 +287,13 @@ class goto_programt
       return guard;
     }
 
+    /// Get the condition of gotos, assume, assert
+    exprt &get_condition()
```
Comment (tautschnig) [inline src/goto-programs/goto_program.h]: I believe we do have a lot of places where we have a const and a non-const variant with the same name, but I also think that in those cases we neither have a `get_` nor a `set_` prefix? What I would have used is `void set_condition(exprt cond) { PRECONDITION(has_condition()); guard = std::move(cond); }`, which of course requires a somewhat bigger change in that all users need to do `exprt modified_condition = get_condition(); do_something(modified_condition); set_condition(std::move(modified_condition));` But certainly nothing that I intend to block on, my approval remains in place.

Comment (kroening) [reply src/goto-programs/goto_program.h]: Indeed; I couldn't decide. To write to stuff thats returned by `get_X` seems to be very common in Java, but much less common in C++.
The key reason why I went for non-const get is that we have a lot of instances where we apply a transform to the guard. That would then become
`exprt tmp = instr.get_condition();
TRANSFORM(tmp);
instr.set_condition(tmp);`
That could possibly be changed into
`instr.set_condition(TRANSFORM(instr.get_condition());`
if we turn all those TRANSFORM into functional style.

Comment (kroening) [reply src/goto-programs/goto_program.h]: I'll action this.

Comment (kroening) [reply src/goto-programs/goto_program.h]: Another option would be to offer `instructiont::transform()`, which would take a lambda that modifies the expressions.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: bfb52872e46c1135d13b39427df8bb555d63d96b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100579413
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 09371f1923a822a1cb3ccd12aaab22c0a745e057).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100582757

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d1c27498f13d06fb485e7d091c09472fcb70fb99).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100736589

**Pull request merged**